tests/utils/runner_tests.py

```diff
@@ -0,0 +1,359 @@
+        expected_err: str | re.Pattern,
+        caplog: pytest.LogCaptureFixture,
+    ):
+        if not warn:
+            # No warnings should have been logged.
+            assert not any(
+                record.levelname in ("WARNING", "ERROR", "CRITICAL")
+                for record in caplog.records
+            )
+            return
+
+        if hide is True:
+            # Warnings not logged at all (because `warn=True` and `hide=True`).
+            assert caplog.records == []
+        elif isinstance(expected_err, str):
+            assert len(caplog.records) == 1
+            assert (
+                caplog.records[0].message.strip()
+                == f"Command {command!r} returned non-zero exit code 1: {expected_err}"
+            )
+        elif isinstance(expected_err, re.Pattern):
+            assert len(caplog.records) == 1
+            message = caplog.records[0].message.strip()
+            assert expected_err.search(message)
+
+    @pytest.mark.parametrize("use_async", [False, True], ids=["sync", "async"])
+    @pytest.mark.asyncio
+    async def test_get_output_calls_run(
+        self,
+        runner: Runner,
+        use_async: bool,
+        monkeypatch: pytest.MonkeyPatch,
+    ):
+        mock = Mock(spec=subprocess.CompletedProcess, stdout=Mock())
+        command = "echo OK"
+        if use_async:
+            if (
+                runner.run_async is type(runner).run_async
+                and runner.get_output_async is type(runner).get_output_async
+            ):
+                # It's a static method! Path the class instead of the "instance".
+                monkeypatch.setattr(
+                    type(runner),
+                    type(runner).run_async.__name__,
+                    AsyncMock(spec=runner.run_async, spec_set=True, return_value=mock),
+                )
+            else:
+                monkeypatch.setattr(
+                    runner,
+                    runner.run_async.__name__,
+                    AsyncMock(spec=runner.run_async, spec_set=True, return_value=mock),
+                )
+            output = await runner.get_output_async(command)
+        else:
+            if (
+                runner.run is type(runner).run
+                and runner.get_output is type(runner).get_output
+            ):
+                # It's a static method! Path the class instead.
+                monkeypatch.setattr(
+                    type(runner),
+                    type(runner).run.__name__,
+                    Mock(spec=runner.run, spec_set=True, return_value=mock),
+                )
+            else:
+                # It's a regular method:
+                monkeypatch.setattr(
+                    runner,
+                    runner.run.__name__,
+                    Mock(spec=runner.run, spec_set=True, return_value=mock),
+                )
+            output = runner.get_output(command)
+        assert isinstance(output, Mock)
+        assert output is mock.stdout.strip()
```

Comment:
`mock.stdout.strip()` is actually a Mock (since `mock` is a Mock, `mock.stdout` is a Mock, and `mock.stdout.strip` is a Mock, and `mock.stdout.strip()` returns a Mock  - lots of mocks!